lib/src/model/notifications/local_notification_service.dart

```diff
@@ -0,0 +1,156 @@
+
+@Riverpod(keepAlive: true)
+LocalNotificationService localNotificationService(
+  LocalNotificationServiceRef ref,
+) {
+  return LocalNotificationService(ref, Logger('LocalNotificationService'));
+}
+
+class LocalNotificationService {
+  LocalNotificationService(this._ref, this._log);
+
+  final LocalNotificationServiceRef _ref;
+  final Logger _log;
+  final Map<int, void Function(String?, String?)> _callbacks = {};
+  final _notificationPlugin = FlutterLocalNotificationsPlugin();
+  final _receivePort = ReceivePort();
+
+  int currentId = 0;
+
+  Future<void> init() async {
+    _updateLocalisations();
+    _ref.listen(generalPreferencesProvider, (prev, now) {
+      if (prev!.locale == now.locale) return;
+      _updateLocalisations();
+    });
+
+    // hot reloading doesnt remove the port so we just make sure it doesnt exist before registering it
+    IsolateNameServer.removePortNameMapping('localNotificationServicePort');
+    IsolateNameServer.registerPortWithName(
+      _receivePort.sendPort,
+      'localNotificationServicePort',
+    );
+    _receivePort.listen(_onReceivePortMessage);
+
+    await _notificationPlugin.initialize(
+      InitializationSettings(
+        android: const AndroidInitializationSettings('logo_black'),
+        iOS: DarwinInitializationSettings(
+          requestAlertPermission: false,
+          requestBadgePermission: false,
+          requestSoundPermission: false,
+          notificationCategories: <DarwinNotificationCategory>[
+            ChallengeNotificationDetails.instance.darwinNotificationCategory,
+          ],
+        ),
+      ),
+      onDidReceiveNotificationResponse: _notificationResponse,
+      onDidReceiveBackgroundNotificationResponse:
+          _notificationBackgroundResponse,
+    );
+    _log.info('[Local Notifications] initialized');
+  }
+
+  void _updateLocalisations() {
+    final l10n = _ref.read(l10nProvider);
+    InfoNotificationDetails(l10n.strings);
+    ChallengeNotificationDetails(l10n.strings);
+  }
+
+  Future<int> show(LocalNotification notification) async {
+    final id = currentId++;
+    await _notificationPlugin.show(
+      id,
+      notification.title,
+      notification.body,
+      notification.notificationDetails,
+      payload: notification.payload,
+    );
+    _log.info(
+      '[Local Notifications] Sent notification: ($id | ${notification.title}) ${notification.body} (Payload: ${notification.payload})',
+    );
+
+    // if the notification implements LocalNotificaitonCallback
+    if (notification is LocalNotificationCallback) {
+      _callbacks[id] = (notification as LocalNotificationCallback).callback;
+      _log.info('[Local Notifications] registered callback for id [$id]');
+    }
+
+    return id;
+  }
+
+  Future<void> cancel(int id) async {
+    _log.info('[Local Notifications] canceled notification id: [$id]');
+    return _notificationPlugin.cancel(id);
+  }
+
+  void _onReceivePortMessage(dynamic message) {
```

Comment:
I don't understand why it is still needed to send message to the main isolate. Can you explain why it is useful? Thanks.

I still feel it should not be necessary, but maybe I'm missing sth.

Reply:
the thing is we dont have much information about the notification, what you suggest below with directly sending a message to the api in _notificationBackgroundResponse() is definitely possible, but this is a very basic example we have here with only one type of notification.

with multiple notification types that have to do some action in the background isolate the only real meaningful data that we have access to is the actionId and the payload.

this means that we will have to somehow encode the type of notification that is being interacted with in one of these two things (probably the payload) which could look something like
"{notificationType}:{payload}"
as soon as we start dealing with 4 or 5 different types of notifications itll get pretty messy

sending messages between isolates also makes defining new notifications easier as all the behaviour associated with that notification is declared in that one function, instead of two different places which i think improves the code readability.

i dont think its impossible to have it separate and to keep it isolated, but i also dont understand the tradeoff. it seems like forwarding the notificationResponse to the main isolate to handle just makes things simpler and more readable without any downside.

Reply:
When threading is involved I'm not sure things are more readable (or understandable) but I get your point.

When your speaking about multiple notification type your thinking about future usage? But we don't know yet what that will be yet.

We can define an interface for the payload, so that we always know the type of notification:
```dart
class NotificationPayload<T> {
   final String notifType;
   final T data;
}
```

Used in all notifications, (de)serialised to JSON. That's also how it is done with push notifications.
   
Also there's another thing that I'm not sure about. When you receive a notification when the app is in background there are 2 possibilities:
- user declines: then the app should stay in background and the cancel request should be send to server
- user accepts: app should be bring to foreground

I was seeing things like that, but maybe the flutter local notification plugin doesn't work that way, you'll tell me:
- in the decline case we have to deal with the limited background isolate where we don't have access to the whole app but that's ok if we only need to send a request.
- in the accept case, I'd expected a similar mechanism as the [handleMessage callback](https://github.com/lichess-org/mobile/blob/af96a89efb6cbb1e0ea67b3c20845b262a32bd6f/lib/src/app.dart#L374) of the firebase push plugin (that fires when a message clicked caused the app to open). Do we have a similar thing in the local notification plugin?

Reply:
>When threading is involved I'm not sure things are more readable (or understandable) but I get your point.

i really think it does.
now the behaviour relating to the challenge notification is split up into 3 different files, and there is a bunch of casting and null checks to extract data from the notification payload.
i think that with your solution of keeping things separate it becomes way more confusing as now you actually **do** have to think about different threads as with some reponses youre in a background isolate and with some you arent.

not only that but as seen in this https://github.com/lichess-org/mobile/pull/902#discussion_r1711238492 we dont have access to any refs, so now a static instance is required somewhere to get access to one as doing anything with the lichess servers requires a LichessClient, which in and of itself requires a ref.

with messaging the main isolate everything is just abstracted away and in future it doesnt have to be thought about again. maybe the current implementation could be a little cleaner, maybe by turning the NotificationReponse into json and then sending over the RecievePort and deserializing it there.
this means that all the behaviour relating to a notification is just included in the same file it's declared in and when implementing it you dont need to think about whether or not a particular response will open the application or not because it has already been handled for you.

> I was seeing things like that, but maybe the flutter local notification plugin doesn't work that way, you'll tell me:

essentially how the flutter_local_notfication package works is that there are two different types of interactions a user can have with a notification. One that opens the application and one that doesnt.

in the case that the application is opened _notificationReponse is called,
and in the case that the applciation remains closed _notificationBackgroundResponse is executed.

we can handle it similarly to how handleMessage does with firebase notifications as seen below, but this function is going to get very big very quickly.

```dart

  /// Handle a notification response that caused the application to open
  Future<void> _notificationResponse(NotificationResponse response) async {
    if (response.payload == null) return;
    final payload = NotificationPayload.fromJson(
      jsonDecode(response.payload!) as Map<String, Object?>,
    );

    switch (payload.type) {
      case NotificationType.info:
      case NotificationType.challenge:
        switch (response.actionId) {
          case 'accept':
            final repo = ref.read(challengeRepositoryProvider);
            final id = payload.data['id'] as String;
            final challengeId = ChallengeId(id);

            await repo.accept(challengeId);
            final fullId = await repo
                .show(challengeId)
                .then((challenge) => challenge.gameFullId);

            _popNavState();
            pushPlatformRoute(
              ref.read(currentNavigatorKeyProvider).currentContext!,
              rootNavigator: true,
              builder: (BuildContext context) =>
                  GameScreen(initialGameId: fullId),
            );
          case null:
            _popNavState();
            pushPlatformRoute(
              context,
              builder: (BuildContext context) =>
                  const ChallengeRequestsScreen(),
            );
        }
    }
  }
```

>When your speaking about multiple notification type your thinking about future usage? But we don't know yet what that will be yet.

i have already started working on #459 and ill probably add notifications to that as well as the ability to respond to a message within the notification itself.
basically i dont think it matters what type of notifications are added in the future. As long as there are more notifications in the future and we handle them like this the _notificationReponse functions will become more messy regardless.

Reply:
Ok, let's keep for now the message to main isolate, I prefer to focus on the more important matters. A couple of questions about it though:
- what happens if the use kills the app, then for instance, cancels a challenge through a notification? does the notification disappear with the app being killed? or is the message just being ignored?
- a scheduled notification cannot send messages to main isolate because we cannot assume the app will be launched when the notification action is tapped.

Now, let's deal with the notifications handlers. Here I'm sure that I don't want the callback based system. My main concern is code readability and maintenance over years. With callbacks it is definitely harder to understand, especially if you come back to the code months after, or for anybody not familiar with the codebase.

So we definitely need a payload type as I described in my other comment:
```dart
class NotificationPayload<T> {
   final PayloadType payloadType;
   final T data;
}
```

It has the added benefit of a common interface for all notification messages the app will have to deal with in the future. And it adds another type safety layer, over just using `String` and let the message handler define their own format.

> we can handle it similarly to how handleMessage does with firebase notifications as seen below, but this function is going to get very big very quickly.

I won't be big if we use the following approach:
```dart
  /// Handle a notification response that caused the application to open
  Future<void> _notificationResponse(NotificationResponse response) async {
    if (response.payload == null) return;
    final payload = NotificationPayload.fromJson(
      jsonDecode(response.payload!) as Map<String, Object?>,
    );

    switch (payload.type) {
      case PayloadType.challenge:
         ref.read(challengeService).onNotificationResponse(response, payload);
      case PayloadType.message:
         ref.read(messagingService).onNotificationResponse(response, payload);
    }
  }
```

That way we still have a clean separation of concerns: the notification service only job is to receive and dispatch the notifications to the proper handlers. It is not messy and it is easy to follow. All message types that the app deal with are defined in only one place.